plugin/trino-sqlserver/src/main/java/io/trino/plugin/sqlserver/SqlServerClient.java

```diff
@@ -768,6 +774,25 @@ private static SliceWriteFunction nvarcharWriteFunction()
                 .flatMap(dataCompression -> Enums.getIfPresent(DataCompression.class, dataCompression).toJavaUtil());
     }
 
+    private static Optional<DataCompression> getTableDataCompressionWithRetries(Handle handle, JdbcTableHandle table)
+    {
+        // DDL operations can take out locks against system tables causing the `getTableDataCompression` query to deadlock
+        // Retry if this deadlock occurs
+        RetryPolicy<Optional<DataCompression>> retryPolicy = new RetryPolicy<Optional<DataCompression>>()
+                .withMaxAttempts(3)
+                .handleIf(throwable ->
+                {
+                    Throwable rootCause = Throwables.getRootCause(throwable);
+                    return com.microsoft.sqlserver.jdbc.SQLServerException.class.isAssignableFrom(rootCause.getClass()) &&
+                           rootCause.getMessage().contains("was deadlocked on lock resources with another process");
```

Comment:
SQLServerException class has `getSQLServerError` method which returns `SQLServerError` class that has `getErrorNumber` method. We should assert on that instead of an error message.

Reply:
We could also run this query with `WITH (UPDLOCK)` hint.

Reply:
Fixed the first comment, about the second:
After reading the docs, I'm not sure if that hint would work on system tables but I am by no means very familiar with MSSQLServer. 

Even if the updlock hint works on system tables, would that block schema migrations from happening when this query was running? It might make more sense to let the schema migrations happen and then retry the (relatively cheaper) deadlocked listing of tables.